components/binding-usage-controller/internal/controller/usagekind/protection_controller.go

```diff
@@ -0,0 +1,304 @@
+
+	usageKindLister       ukLister.UsageKindLister
+	usageKindListerSynced cache.InformerSynced
+	ukClient              ukClient.ServicecatalogV1alpha1Interface
+
+	log        logrus.FieldLogger
+	maxRetires int
+
+	// test hooks used only in unit tests
+	testHookAddFinalizerDone    func()
+	testHookProcessDeletionDone func()
+}
+
+type namespaceNameKey struct {
+	Namespace string
+	Name      string
+}
+
+// NewProtectionController creates a new instance of ProtectionController.
+func NewProtectionController(
+	kindInformer ukInformer.UsageKindInformer,
+	sbuInformer sbuInformer.ServiceBindingUsageInformer,
+	usageKindInterface ukClient.ServicecatalogV1alpha1Interface,
+	log logrus.FieldLogger,
+) *ProtectionController {
+	serviceBindingUsageInformer := sbuInformer.Informer()
+	serviceBindingUsageInformer.AddIndexers(cache.Indexers{
+		indexKind: func(obj interface{}) ([]string, error) {
+			sbu, ok := obj.(*v1alpha1.ServiceBindingUsage)
+			if !ok {
+				return nil, fmt.Errorf("cannot convert item, expected ServiceBindingUsage, but was %T", obj)
+			}
+			return []string{sbu.Spec.UsedBy.Kind}, nil
+		},
+	})
+
+	c := &ProtectionController{
+		queue:                    workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "UsageKindProtection"),
+		bindingUsageInformer:     serviceBindingUsageInformer,
+		bindingUsageListerSynced: serviceBindingUsageInformer.HasSynced,
+		usageKindLister:          kindInformer.Lister(),
+		usageKindListerSynced:    kindInformer.Informer().HasSynced,
+		ukClient:                 usageKindInterface,
+
+		log: log.WithField("service", "controller:usage-kind-protection"),
+
+		maxRetires: defaultMaxRetries,
+	}
+
+	kindInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: c.onAddOrUpdateUsageKind,
+		UpdateFunc: func(old, new interface{}) {
+			c.onAddOrUpdateUsageKind(new)
+		},
+		// no need to react on deletion
+	})
+
+	// TODO: UsageKind must not be deleted before referenced ServiceBindingUsage deletion work is done
+	// TODO: after adding finalizer to SBU or other solution, uncomment this
+	//sbuInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+	//	DeleteFunc: c.OnDeleteSBU,
+	//})
+	return c
+}
+
+func (c *ProtectionController) onAddOrUpdateUsageKind(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		c.log.Errorf("while handling addition event: couldn't get key: %v", err)
+		return
+	}
+	c.queue.Add(key)
+}
+
+// OnDeleteSBU reacts on ServiceBindingUsage deletion
+func (c *ProtectionController) OnDeleteSBU(event *controller.SBUDeletedEvent) {
+	// TODO: this is temporary solution, the method should be removed when SBU informer is implemented
+	c.log.Debugf("OnDeleteSBU %s/%s", event.Namespace, event.Name)
+
+	// make processing asynchronously
+	go func() {
+		uk, err := c.usageKindLister.Get(event.UsedByKind)
+		if err != nil {
+			c.queue.AddRateLimited(event.UsedByKind)
```

Comment:
It is not a infinite loop. In case of error, the kind is put into the working queue. There is a check in processNextWorkItem with max retries. I've added check if it is NotFoundError - such error is ignored - no need any processing.